crates/bitwarden-ssh/src/generator.rs

```diff
@@ -0,0 +1,88 @@
+use serde::{Deserialize, Serialize};
+use ssh_key::{rand_core::CryptoRngCore, Algorithm, HashAlg, LineEnding};
+#[cfg(feature = "wasm")]
+use tsify_next::Tsify;
+
+use crate::{error, error::KeyGenerationError, SshKey};
+
+#[derive(Serialize, Deserialize)]
+#[cfg_attr(feature = "wasm", derive(Tsify), tsify(into_wasm_abi, from_wasm_abi))]
+pub enum KeyAlgorithm {
+    Ed25519,
+    Rsa3072,
+    Rsa4096,
+}
+
+pub fn generate_sshkey(key_algorithm: KeyAlgorithm) -> Result<SshKey, error::KeyGenerationError> {
+    let rng = rand::thread_rng();
+    generate_sshkey_internal(key_algorithm, rng)
+}
+
+fn generate_sshkey_internal(
+    key_algorithm: KeyAlgorithm,
+    mut rng: impl CryptoRngCore,
+) -> Result<SshKey, error::KeyGenerationError> {
+    let key = match key_algorithm {
+        KeyAlgorithm::Ed25519 => ssh_key::PrivateKey::random(&mut rng, Algorithm::Ed25519),
+        KeyAlgorithm::Rsa3072 | KeyAlgorithm::Rsa4096 => {
+            let bits = match key_algorithm {
+                KeyAlgorithm::Rsa3072 => 3072,
+                KeyAlgorithm::Rsa4096 => 4096,
+                _ => unreachable!(),
+            };
+
+            let rsa_keypair = ssh_key::private::RsaKeypair::random(&mut rng, bits)
+                .map_err(KeyGenerationError::KeyGenerationError)?;
+
+            let private_key =
+                ssh_key::PrivateKey::new(ssh_key::private::KeypairData::from(rsa_keypair), "")
+                    .map_err(KeyGenerationError::KeyGenerationError)?;
+            Ok(private_key)
```

Comment:
🎨 Seems like the duplication in the match arms could be eliminated by making these lines into a function.

```rust
// add the appropriate types here
fn create_rsa_key(&mut rng, bits) {
   let rsa_keypair = ssh_key::private::RsaKeypair::random(&mut rng, bits)
       .map_err(KeyGenerationError::KeyGenerationError)?;
    let private_key = ssh_key::PrivateKey::new(ssh_key::private::KeypairData::from(rsa_keypair), "")
        .map_err(KeyGenerationError::KeyGenerationError)?;
    Ok(private_key)
}
```

```rust
let key = match key_algorithm {
    KeyAlgorithm::Ed25519 => ssh_key::PrivateKey::random(&mut rng, Algorithm::Ed25519),
    KeyAlgorithm::Rsa3072 => create_rsa_key(3072),
    KeyAlgorithm::Rsa4096 => create_rsa_key(4096),
}
```

🎨 This could be simplified further by adding the key length as an algorithm parameter, though that might not be appropriate since only specific key lengths are allowed.

```rust
pub enum KeyAlgorithm {
    Ed25519,
    Rsa(usize),
}
```